### ISSUE TITLE
Query: CloughTocher2DInterpolator  reproducibility with rectangular grid

### ISSUE THREAD

**Issue by tylerjereddy** [scipy/scipy#21804]
We have an internal use case for `CloughTocher2DInterpolator` with a regular grid where `RegularGridInterpolator` is taking much longer to run, so the user would really like to be able to use `CloughTocher2DInterpolator` with the regular/rectangular grid reproducibly.

The problem, of course, is that a regular grid violates general position, so the Delaunay triangulation used under the hood by `CloughTocher2DInterpolator` is not guaranteed to be unique, and so i.e., our testsuite ceases to be portable on different supercomputers and so on.

I was able to reproduce a pretty substantial difference in the interpolation output for the code below the fold when just running it on i.e., my mac locally vs. a Linux box (with latest SciPy `main`).

<details>

```python
import numpy as np
import scipy
from scipy.interpolate import CloughTocher2DInterpolator

x = np.array([0, 1, 2, 3, 4, 5, 6, 7]) 
y = np.array([0, 1, 2, 3, 4, 5, 6, 7]) 
X, Y = np.meshgrid(x, y, indexing='ij')
points = np.column_stack([X.ravel(), Y.ravel()])
z = np.sqrt(points[..., 0] + points[..., 1])
X1 = np.linspace(min(x), max(x), num=500)
Y1 = np.linspace(min(y), max(y), num=500)
X1, Y1 = np.meshgrid(X1, Y1)
interp = CloughTocher2DInterpolator(points, z)
Z = interp(X1, Y1)
print("Z:", Z.sum(), Z.min(), Z.max())
```

</details>

- mac: `Z: 644349.4631240882 0.0 3.7416573867739413`
- Linux: `Z: 644324.0908721345 0.0 3.7416573867739413`
(i.e, the first printed value)

If I manually joggle the generators via Qhull, `tri = Delaunay(points, qhull_options="QJ")`, the results appear to be *extremely* close:

- mac: `Z: 644552.4920765815 0.0 3.7416573867739413`
- Linux: `Z: 644552.4920765815 0.0 3.7416573867739413`

The user did indicate that they tried to distort the grid a bit themselves, and it did work sometimes, but depending on the grid density it sometimes was not sufficient for their purposes. I'll point them to this issue to see if this joggling approach suffices, but if not I know they had an interest in being able to provide a custom triangulation of the rectangular grid to `CloughTocher2DInterpolator`. Anyway, I guess let's first see if the joggling suffices for them, and if not I'd be curious to hear suggestions from the `interpolate` folks.

**Comment by ev-br** [comment]
> case for CloughTocher2DInterpolator with a regular grid where RegularGridInterpolator is taking much longer to run

Some details would be great. Or even a runnable example to demonstrate the difference. Maybe there's something to improve in RGI

**Comment by tylerjereddy** [comment]
It sounds like the user was able to solve the immediate issue by switching to `RectBivariateSpline` (fast + rectangular support).

From their feedback, it sounds like the main thing they'd be interested in longer term is providing a custom triangulation to Clough-Tocher. The specific example they mentioned was https://matplotlib.org/stable/api/tri_api.html, which supports interpolation on a custom triangular grid, but suffers substantial performance issues with the underlying `CubicTriInterpolator` performance being fairly slow (3-4x slower than a typical SciPy interpolator it seems).

**Comment by ev-br** [comment]
Now that I actually read the docs, https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.CloughTocher2DInterpolator.html, it appears that my earlier "optional kwarg" comment was noise: Clough-Tocher accepts a precomputed Delaunay object:
https://github.com/scipy/scipy/blob/main/scipy/interpolate/_interpnd.pyx#L66

Which begs the question if this is enough for the user's long-term interest; and if not, then why not.

**Comment by ev-br** [comment]
Also if `RegularGridInterpolator` is too slow and the grid is actually uniform (image data?), it might make sense for them to try something along the lines of https://docs.scipy.org/doc/scipy/tutorial/interpolate/ND_regular_grid.html#uniformly-spaced-data

**Comment by tylerjereddy** [comment]
> Which begs the question if this is enough for the user's long-term interest; and if not, then why not.

I guess that's the original comment I made--the generators are not in general position so Delaunay isn't deterministic. But yeah, both you and I have provided different kinds of workarounds for that scenario now, so I'll close this I think.

If I'm provided with an in-house reproducer I'll consider reopening after I "anonymize" it, but have not been provided with one.
